File: programmers/코딩테스트 입문 with Java/다항식더하기.java
package 코딩테스트입문;

/**
 * https://school.programmers.co.kr/learn/courses/30/lessons/120863?language=java
 */
public class 다항식더하기 {
    public static void main(String[] args) {
        System.out.println(solution("3x + 7 + x"));
        System.out.println(solution("x + x + x"));
        System.out.println(solution("x"));
        System.out.println(solution("9"));
        System.out.println(solution("3x"));
    }

    public static String solution(String polynomial) {
        int num1 = 0;
        int num2 = 0;
        StringBuilder sb = new StringBuilder();
        for (String str : polynomial.split(" \\+ ")) {
            if (str.contains("x")) {
                str = str.replace("x", "");
                if (str.equals("")) {
                    num1++;
                } else {
                    num1 += Integer.parseInt(str);
                }
            } else {
                num2 += Integer.parseInt(str);
            }
        }

        if (num1 > 1) sb.append(num1);
        if (num1 != 0) sb.append("x");
        if (num1 != 0 && num2 != 0) sb.append(" + ");
        if (num2 != 0) sb.append(num2);

        return sb.toString();
    }
}

File: programmers/코딩테스트 입문 with Python/다항식 더하기.py
def solution(polynomial):
    my_list = polynomial.split(" ")
    my_list.sort()
    a, b = 0, 0
    for l in my_list:
        if l == '+':
            continue
        if l.count('x'):
            if l == 'x':
                a += 1
            else:
                l = l.replace('x', '')
                a += int(l)
        else:
            b += int(l)

    answer = 'x'

    if a > 1 and b > 0:
        answer = str(a) + 'x + ' + str(b)
    elif a == 1 and b > 0:
        answer = 'x + ' + str(b)
    elif a > 1 and b == 0:
        answer = str(a) + 'x'
    elif a == 0:
        answer = str(b)

    return answer


print(solution("3x + 7 + x"))
print(solution("x + x + x"))
print(solution("2x"))
print(solution("x + 2"))
print(solution("2 + x"))
print(solution("1"))
print(solution("2"))
print(solution("2 + 2"))
print(solution("2 + 2 + 2x"))
print(solution("x + 2x"))
print(solution("x"))
print(solution("x + 1"))
print(solution("x + 1"))
print(solution("1 + x + 1"))
print(solution("x + x + 1"))
print(solution("9 + 2x"))
print(solution("2x + 2x + 0"))
print(solution("0 + 2x + 0"))
print(solution("0 + 0"))
print(solution("0"))
print(solution("0 + 2 + 9"))
print(solution("0x + 2 + 9"))
print(solution("0x + 0x + 2x"))




